README.txt
 ____________________________________________________
|                                                    |
|            *** DNBNOR-TO-QIF ***                   |
|____________________________________________________|

 dnbnor2qif is a simple tool to help integrate data from the 
   DnBNOR online bank monthly transcripts ("kontoutskrift")
     to a QIF accepting financial program, i.e. GnuCash.

                  Kent Dahl 
            <kentda AT pvv DOT org>

              Copyright 2007-2008
         Released under the Ruby license.

Available from:
          http://dnbnor2qif.rubyforge.org/
          http://rubyforge.org/projects/dnbnor2qif/
          http://www.pvv.org/~kentda/ruby/
 
____________________________________________________
Disclaimer:

This is an experimental hack and you may end up bankrupt or worse
if you try to rely on its correct functioning. You have been warned.

This program is in no way, what so ever, implied or otherwise, endorsed by
nor affiliated with DnB NOR (https://www.dnbnor.no/). You should really go
bug them about providing OFX downloads directly instead.

____________________________________________________
Contents of this document:

- Requirements
- How to use
- Plans for the future
- Related links
- Other alternatives

____________________________________________________
Requirements:

* Ruby 1.8.x or newer
  -- http://www.ruby-lang.org/

* roo 1.0.1 or newer
  -- http://roo.rubyforge.org/
  -- To install: gem install roo


____________________________________________________
How to use dnbnor2qif:

First download the bank statement from the online bank site:
- Click on "Totaloversikt" and select the account.
- Click on the "Kontoutskrift" tab and fetch the month you want.
- Right-click on "Kopier til regneark" link and save the document.

Then run the script from the command line:
  ruby dnbnor2qif.rb -c [command] -i [input] [ -o [output] ]

Options:
  -c, --command   |- Command to execute, such as 'convert'.
  -i, --input     |- Input file (CSV from DnBNOR)
  -o, --output    |- Output filename (QIF) (optional)
  -a, --account   |- Full name of GnuCash bank account (optional)
  -d, --debug     |- Verbose debugging output (experimental)


____________________________________________________
Plans for the future:

See the TODO.txt file.


____________________________________________________
Related links:

* expenses2qif for the Agenda VR3
  -- http://agtoys.sourceforge.net/ruby/

* qif/qif.rb
  -- very simple QIF writer.

____________________________________________________
Other potential alternatives:

* MT2OFX
  -- http://www.xs4all.nl/~csmale/mt2ofx/en/

* XL2QIF
  -- http://xl2qif.chez-alice.fr/xl2qif_en.php

* QIF Master 
  -- (for MacOS X only)
  -- http://www.thewoodwards.us/sw/QIFMaster/

* MnyBank
  -- http://ferraroa.dyndns.org:8200/MnyBank/help/doc.htm

____________________________________________________


NEWS.txt
Very preliminary support for Sparebanken Pluss CSV format.
Preliminary support for Postbanken format.
- Postbanken is based on DnB NOR.


test/run.rb
require 'test/test_qif'
require 'test/test_dnbnor'
require 'test/test_parser'


test/test_qif.rb
require 'test/unit'
require 'qif/qif'

require 'stringio'

include Qif

class TestAccount < Test::Unit::TestCase
  def setup
    @io = StringIO.new
    @cash = Qif::Account.new( 'Wallet', nil, 'My cash' ) 
  end

  def test_account_output
    @cash.output_to( @io )
    assert_equal "!Account\nNWallet\nDMy cash\n^\n", @io.string
  end

  def test_transaction_output
    @food = Qif::Account.new( 'Food',   'Grub'    )
    t = Qif::Type::new('Cash')
    t.date = Time.at( 1059429288 )
    t.debit_acc = @food
    t.credit_acc = @cash
    t.output_to( @io )
    assert_equal "!Type:Cash\nD07-28-2003\n^\n", @io.string
  end

end




class TestQIFType < Test::Unit::TestCase
  def setup
    @cash = Qif::Type.new("Cash") 
    @bank = Qif::Type.new("Bank")
    @card = Qif::Type.new("CCard") 

  end

  def test_type_new
    assert( @cash )
    assert( @bank )
    assert( @card )
  end

  def test_type_fail
    assert_raises( ArgumentError ){ Qif::Type.new("Dinero") }
    assert_raises( ArgumentError ){ Qif::Type.new("Money") }
    assert_raises( ArgumentError ){ Qif::Type.new("Muchalucha") }
  end

end


test/test_dnbnor.rb
=begin

Test cases for DnBNOR (tm) data file parsing.

=end


require 'test/unit'
require 'dnbnor/dnbnor.rb'

class TestDnBNOR < Test::Unit::TestCase
  def setup
    @entry = DnBNOR::Entry.new
  end

  def test_parse_value_date
    assert_equal( "1978-05-25", @entry.parse_value(:date, "25.05.78").to_s)
    assert_equal( "2007-06-05", @entry.parse_value(:date, "05.06.07").to_s)
    assert_equal( "2000-12-31", @entry.parse_value(:interest_date, "31.12.00").to_s)
    assert_equal( nil, @entry.parse_value(:interest_date, nil))

  end

  def test_parse_value_amount
    assert_equal(  256.99, @entry.parse_value(:credit, "256,99") )
    assert_equal(  512.78, @entry.parse_value(:debit,  "512.78") )
    assert_equal( 1512.78, @entry.parse_value(:debit,  "1512.78") )
  end


  DETAILS_TC1 = "Varekjøp  55551111               ,Ark.ref *17171717 Dato 08.08 Kl. 11.19 Versjon 1 Aut. 12345     Kiwi Sannergata Sannergt.3 Oslo"
  DETAILS_TC2 = "Visa  112233                       ,Kr.sand Parkeringsselskap        "
  def test_parse_details

    c = @entry.parse_details( DETAILS_TC1 )
    assert_equal( :purchase,  c[:type] )
    assert_equal( 55551111,  c[:transaction_id])
    assert_equal( 17171717,  c[:archive_id])
    assert_equal( 12345,     c[:authentication_id])
    assert_equal( "08.08",   c[:date])
    assert_equal( "11:19",   c[:time])
    assert_equal( "Kiwi Sannergata Sannergt.3 Oslo",   c[:comment])

    c = @entry.parse_details( DETAILS_TC2 )
    assert_equal( :visa,  c[:type] )
    assert_equal( "Kr.sand Parkeringsselskap (Visa)",   c[:comment])

  end


  def test_parse_no_description
    c = @entry.parse_details( nil )
    assert_not_nil c
  end

  def test_empty_entry
    assert_equal( true, @entry.empty? )

    @entry['Dato'] =  nil
    assert_equal( true, @entry.empty? )

  end

end


class TestPostbanken < Test::Unit::TestCase
  def setup
    @entry = DnBNOR::Entry.new
  end


  def test_posbanken_payment_number
    c = @entry.parse_details("FAST OPPDR. FOND POSTBANKEN NORGE        Betnr:  561     ")
    assert_equal( 561, c[:archive_id] )
    assert_equal( "FAST OPPDR. FOND POSTBANKEN NORGE", c[:comment] )

    c = @entry.parse_details("AVTALEGIRO       NextGenTel AS           BETNR: 0562     ")
    assert_equal( 562, c[:archive_id] )
    assert_equal( "AVTALEGIRO       NextGenTel AS", c[:comment] )

  end


end





qif/qif.rb
=begin
QIF - Quicken Interchange Format is an old, stale, de facto standard.

So we are just writing up the bits we need for importing the 
expenses.dat-file into other programs.

=end

##
#
module Qif
  

  ##
  #
  module QifEntry
    
    attr_accessor :name
    
    DEFAULT_FIELD_TYPES = 
      [
       # variable name, field code and formatting method
       [:@name, 'N'],
       [:@memo, 'M'],
       [:@description, 'D'],
       [:@date, 'D', :format_date],
       # [:@credit_acc, 'L'],
       [:@category,  'L'],
       [:@amount, 'T'],
       [:@account_type, 'T'],
       [:@payee, 'P'],
       [:@address, 'A', :format_address],       
       [:@cleared, 'C'],
       [:@num,     'N'],
    ]

    def field_types
      DEFAULT_FIELD_TYPES
    end


    ##
    # outputs the entry to the given IO.
    def output_to( o )
      o.puts "!#{header_name}"
      each_entry_type do |type|
        var_name, field, formatter = *type
        data = instance_variable_get( var_name )
        if data
          formatted_data = if formatter
                             self.send(formatter, data)
                           else
                             data
                           end
          o.puts "#{field}#{formatted_data}"
        end        
      end
      o.puts "^"
    end

    ## 
    # iterate through the entry fields applicable.
    def each_entry_type( &block )
      field_types.each &block
    end

    def header_name
      self.class.to_s.split(":").last
    end

    ##
    # Formats the date field for output to QIF.
    def format_date( date )
      sprintf "%02d-%02d-%04d", date.month, date.mday, date.year
    end

    ##
    # Formatting hack for address. 
    # Needed to wrap an array of address elements onto several lines.
    def format_address( address )
      return address.to_str if address.respond_to? :to_str
      address.join("\nA")
    end




  end

  class Account
    include QifEntry

    attr_reader :name, :type, :description
    def initialize(name, type = nil, description=nil)
      @name = name
      @type = type
      @description = description
    end    

    def to_s
      @name
    end
  end


  ##
  # The various types of transactions all go into Type.
  class Type
    include QifEntry

    QIF_TYPES = %w{Bank Cash CCard Invst}
    # Others: Oth A, Oth L, Cat, Class, Memorized 

    attr_accessor :date, :amount, :memo, :payee, :address, :num
    attr_accessor :debit_acc, :credit_acc

    ##
    # Valid type names.
    def initialize( type_name )
      QIF_TYPES.include?( type_name) or 
        raise ArgumentError, "Unsupported QIF type: #{type_name}."
      @type_name = type_name
    end

    attr_accessor :type_name

    def header_name
      "Type:#{@type_name}"
    end

  end

end



if __FILE__ == $0 then
  cash = Qif::Account.new( 'Wallet', 'My cash' ) 
  food = Qif::Account.new( 'Food',   'Grub'    )


  t = Qif::Type::new('Cash')
  t.date = Time.now
  t.debit_acc = food
  t.credit_acc = cash


  [cash,food,t].each{|i|
    i.output_to( $stdout )
  }
  

end

=begin
Sample QIF:


-----8<------
!Account
NEiendeler:Kontanter
DKontanter
^
!Type:Cash
D 07-28-2003
T -1024.59
M Legesjekk
L Buss
^
-----8<------
!Account
NUtgifter:Buss
^
!Type:Cash
D 07-28-2003
T 485
M Busskort
LEiendeler:Kontanter
^
-----8<------

=end


dnbnor/dnbnor.rb
=begin

DnBNOR simple accounting data.

=end


require 'date'

module DnBNOR

  DEBUG = false

  ##
  # Single transaction entry from.
  class Entry 

    FIELD_MAP = {
      # Norwegian
      "Dato"         => :date, 
      "Forklaring"   => :details, 
      "Rentedato"    => :interest_date, 
      "Ut fra konto" => :debit,
      "Inn på konto" => :credit,

      # Norwegian variations
      "Ut av konto" => :debit,
      "Inn til konto" => :credit,

      # English
      "Date (mm.dd.yyyy)" => :date, 
      "Description"       => :details, 
      "Interest date"     => :interest_date, 
      "Withdrawals"       => :debit,
      "Deposits"          => :credit,

    }

    AMOUNT_RE = /(\d+[,.]?\d*)/
    DATE_FORMAT = "%d.%m.%y"

    def initialize
      @data = Hash.new
    end

    def []=(index, value)
      key = FIELD_MAP[index]||index
      @data[ key ] = parse_value( key, value )
    end

    def method_missing( sym, *args )
      @data[sym]
    end

    def parse_value( key, value )
      # p [key, value] if DEBUG
      return nil unless value # No value? Nothing we can do.
      case key
      when :date, :interest_date
        return value if value.kind_of?(Date)
        ( value ? Date.strptime(value, DATE_FORMAT) : nil )
      when :debit, :credit
        return value if value.kind_of?(Numeric) 
        ( value =~ AMOUNT_RE ? $1.tr(',','.').to_f : nil )
      when :details
        parse_details( value.chomp.gsub(/\s{2,}/, " ").tr("\n"," ").strip )
      else
        value
      end
    end

    def parse_details( details )
      c = Hash.new

      case details

      when /^Varekj.p\s+(\d+)\s+\,?Ark\.ref\s+\*(\d+)\s+Dato\s+(\d+\.\d+)\s+Kl\.\s+(\d+\.\d+)\s+Versjon\s+\d+\s+Aut\.\s+(\d+)\s+(.*)$/
        c[:type] = :purchase
        c[:date] = $3 # TODO: parse_value( :date, $3 )
        c[:time] = $4.tr('.',':')
        c[:transaction_id] = $1.to_i
        c[:archive_id] = $2.to_i
        c[:authentication_id] = $5.to_i
        c[:comment] = $6

      when /^(L.nn)\s+(\d+)\s+\,?Fra\s+(.*?)\s*$/
        c[:comment] = $1
        c[:archive_id]     = $2.to_i
        c[:payee]   = $3

      when /^Visa\s+(\d+)\s+\,?(.*?)\s*$/
        c[:type] = :visa
        c[:visa] = $1
        c[:comment] = $2 + " (Visa)"        

      when /^(Overføring [InUt]+land)\s+(\d+)\s+(.*?)\s*$/
        c[:archive_id] = $2.to_i
        c[:comment]        = $3 + " (#{$1})"
      when /^(Giro)\s+(\d+)\s+(.*?)\s*$/
        c[:archive_id] = $2.to_i
        c[:comment]        = $3 + " (#{$1})"


      # when /^Kontoregulering/
      # when /^Lønn/
      # when /^Omkostninger/

      when /^(Reservert\s+\-\s+ikke\s+Bokført)\s*$/
        # Should be ignored?
        

      when /^,(.*?)\s*$/
        # Old DnB Nor CSV comment format
        c[:comment] = $1

      when /^\s*(.*?)\s+Betnr\:\s+(\d+)\s*$/i
        # Postbanken, payment number.
        c[:comment] = $1
        c[:archive_id] = $2.to_i

      when /^([^,]+),(.*?)\s*$/
        c[:comment] = "#{$2} (#{$1})"

      else
        c[:comment] = details

      end

      c
    end

    def parse_purchase_details( details )

    end

    def is_debit?
      debit && debit > 0
    end

    def is_credit?
      credit && credit > 0
    end

    def empty?
      @data.empty? || @data.all?{|key,val| val.nil?}
    end

    def to_s
      "<#{@data[:date]} #{@data[:credit]} #{@data[:debit]} \n\t'#{@data[:details].inspect}' >"
    end

  end

end # DnBNOR


dnbnor/dnbnor_parser.rb
#!/usr/bin/ruby

require 'csv'
require 'date'

require 'iconv'

begin
  require 'rubygems'
  gem 'roo', '> 1.0.0'
  require 'roo'
  require 'roo-xls' rescue nil
rescue LoadError
  puts "ERROR: Cannot parse Excel files! Have you got the 'roo' gem?"
  raise
end

require 'dnbnor/dnbnor'

module DnBNOR


  ## 
  # Base class common to parsers.
  # 
  class ParserBase
    attr_reader :entries
    attr_reader :headers

    def initialize
      @entries = Array.new
    end

    def handle_row(row)
      entry = Entry.new
      row.each_with_index{|e,i|
        value = row[i]
        begin
          header = @headers[i]
          entry[@headers[i]] = massage_value(value)
        rescue StandardError => err
          puts "ERROR during parsing of row #{row}, field #{i}(#{header}) containing #{value}."
          p err
          p entry if DEBUG
        end
      }
      @entries << entry unless entry.empty?
    end

    def massage_value(value)
      value
    end


    ##
    # Select parser type based on file extension.
    # 
    def self.parser_type_from_file(input_file)
      case input_file
      when /\.csv$/i
        CSVParser
      when /\.txt$/i
        CSVParser
      when /\.xls$/i
        SpreadsheetParser
      when /\.xl[a-z]+$/i
        SpreadsheetParser
      else
        SpreadsheetParser
      end
    end
    
    ##
    # Create parser based on filename extension
    # and load the file.
    def self.load_from_file(input_file)
      parser_class = self.parser_type_from_file(input_file)
      dnb = parser_class.new
      dnb.load_file(input_file)
      return dnb
    end

  end


  ##
  # Simple DnBNOR transcript parser for the CSV format.
  #
  class CSVParser < ParserBase
    FIELD_SEPARATOR = "\t"

    def load_file( filename )
      File.open( filename, 'r' ){ |file|
        parse_data(file)
      }
    end

    def parse_data( data )               
      csv = CSV::Reader.create( data, "\t" )  # TODO: FIXME!
      @headers = csv.shift
      p headers if DEBUG

      csv.each{|row| handle_row(row) }

      puts @entries if DEBUG      
    end

  end # CSVParser

  ##
  # Default format used to be CSV using tabs
  #
  ## Parser = CSVParser


  ##
  # Alternate DnBNOR transcript parser for the Excel format.
  #

  class SpreadsheetParser < ParserBase

    def load_file( filename )
      @inconv = Iconv.new("ISO-8859-15","UTF-8")

      file = ::Roo::Excel.new(filename)
      file.default_sheet = file.sheets.first # "Kontoutskrift" sheet
      parse_data(file)
    end

    def parse_data( data )
      @data = data
      @headers = data.row(data.first_row).collect{|v| massage_value(v) }

      p @headers if DEBUG

      (data.first_row+1).upto(data.last_row) do |ri|
        handle_row( data.row(ri) )
      end

      @data = nil
      puts @entries if DEBUG      
    end


    def massage_value(value)
      case value
      when String
        value = @inconv.iconv(value)
        value = value.encode("UTF-8")
      end
      value
    end


  end # SpreadsheetParser


  ##
  # Default format now is Excel.
  #
  Parser = SpreadsheetParser




end


util/create_dnbnor2qif_package.sh
#!/bin/bash

BASEPATH=`dirname $0`


##
# Version setup
#
VERSION=`cat $BASEPATH/../VERSION.txt`
FILEVERSION=`echo $VERSION | sed s/\\\\./_/g`
echo $VERSION

TAGPREFIX=dnbnor2qif-alpha-release

PREVTAG=`git-tag | grep $TAGPREFIX | head -n 1`

git-tag -a -m "Tagging release $VERSION"  $TAGPREFIX-$VERSION


##
# Package directory setup
#
TARGETDIR=dnbnor2qif_v${FILEVERSION}
echo $TARGETDIR

rm -rf dist
mkdir -p dist



##
# Changelog 
#
if [ "$PREVTAG" ]
then
    git-log --pretty=format:"%s%n%b" $PREVTAG..$TAGPREFIX-$VERSION \
    | grep -v "^$" \
    > dist/changelog-$VERSION.txt
fi


## 
# Create package
#
echo "Creating archive ..."
TARGETFILE=dist/$TARGETDIR.tar.gz
git-archive --prefix=$TARGETDIR/ master  |gzip>$TARGETFILE
echo "Wrote file: $TARGETFILE".


##
# Run tests.
#
cd dist
tar xzvf `basename $TARGETFILE`
cd $TARGETDIR
echo "Testing..."
ruby test/run.rb 

# Run larger random testsuite.
if [ -d ../../test/various/ ]
then
  for i in ../../test/various/*.xls 
  do  
    ruby dnbnor2qif.rb -c convert -i $i
  done
fi


echo "Done."



dnbnor2qif.rb
#!/usr/bin/ruby
=begin
The command line program.
=end

$LOAD_PATH.push(File.dirname(__FILE__)) # Find my libs

require 'dnbnor/dnbnor_parser'

require 'qif/qif'

require 'iconv'
require 'stringio'


module DnBNOR

  module Converter

    class QIF

      def initialize( args )
        @args = args
        @outconv = Iconv.new("UTF-8","ISO-8859-15")

      end
      
      def command_convert
        input_file = @args['input']
        puts "Converting from file: #{input_file}"
        dnb = DnBNOR::Parser.load_from_file(input_file)
        @entries = dnb.entries

        puts "Loaded #{@entries.size} transactions."


        @bankaccount = Qif::Account.new( @args['account'],
                                         @args['account_type'] )

        @accounts = Hash.new
        @transactions = Array.new

        puts "Converting..."

        convert_entries
        # group_transactions

        puts "Outputing..."
        output_file = @args['output']
        File.open(output_file, 'w' ){|output|
          StringIO.open {|strio|
            output_to( strio )
            output << @outconv.iconv(strio.string)
          }
        }
        
        puts "Outputted to file: #{output_file}"

      end

      def convert_entries
        @entries.each do|e|
          p e if DEBUG

          # Then locate the payment account.
          if e.is_credit? then
            credit_acc = @bankaccount
            debit_acc  = nil
          elsif e.is_debit? then
            debit_acc = @bankaccount
            credit_acc  = nil
          else
            puts "ERROR: Problem with transaction, skipping."
            p e
            next
          end

          # Set up the QIF transaction.
          t = Qif::Type.new("Bank") 
          t.debit_acc  = debit_acc
          t.credit_acc = credit_acc

          # Fairly straight forward element assignment
          # Note that bank statement is inverse of our account view.
          t.amount = (e.is_credit? ? e.credit : -e.debit )
          t.memo   = e.details[:comment]
          t.date   = e.date
          t.num    = (e.details[:archive_id]     || 
                      e.details[:transaction_id] ||
                      e.details[:visa]           || 
                      nil)
          t.payee  = e.details[:payee]

          @transactions << t

        end

        ##
        # Group based on credit (asset) account.
        def group_transactions
          @grouped = Hash.new{ Array.new }
          unless @transactions 
            raise ArgumentError, "No data: Must convert first."
          end
          @transactions.each do |t|
            group = @grouped[ t.debit_acc ]
            group << t
            @grouped[ t.debit_acc] = group
          end
          @grouped
        end


        ##
        # Output QIF
        def output_to( io )
          @bankaccount.output_to( io )
          @transactions.each{|t|
            t.output_to( io )
          }
        end




      end


      USAGE = <<-EOD
Convert DnBNOR CSV to QIF.
Usage:
  ruby dnbnor2qif.rb -c [command] -i [input] [ -o [output] ]
Options:
  -c, --command   |- Command to execute, such as 'convert'.    
  -i, --input     |- Input file (CSV from DnBNOR) 
  -o, --output    |- Output filename (QIF) (optional)
  -a, --account   |- Full name of GnuCash bank account (optional)
  -d, --debug     |- Verbose debugging output (experimental)
EOD
      def command_help
        puts USAGE
      end
      
      
    end # QIF

  end # Converter
  
end # DnBNOR


if __FILE__ == $0 then
  require 'getoptlong'
  
  file = 'kontoutskrift.csv'
  arguments = {'command' => 'help', 
    'input' => file
  }
  opt = GetoptLong.new(
    ['--help',          '-h', GetoptLong::NO_ARGUMENT],
    ['--debug',         '-d', GetoptLong::NO_ARGUMENT],
    ['--input',         '-i', GetoptLong::REQUIRED_ARGUMENT],
    ['--output',        '-o', GetoptLong::REQUIRED_ARGUMENT],
    ['--command',       '-c', GetoptLong::REQUIRED_ARGUMENT],
    ['--account',       '-a', GetoptLong::REQUIRED_ARGUMENT]
  )

  opt.each{|opt,arg|
    s = /\-\-([a-z\-]*)/.match(opt)[1].to_s
    arguments[s] = arg
  }
  
  unless arguments['output'] || !arguments['input'] then
    arguments['output'] ||= arguments['input'] + ".qif"
  end

  # Dummy account name
  arguments['account'] ||= "DnB Nor bank account"

  DnBNOR::DEBUG = true if arguments['debug']

  command = DnBNOR::Converter::QIF.new( arguments )
  command.send( ('command_'+arguments['command']).intern )

end


